cmd/dep/init.go

```diff
@@ -239,6 +242,45 @@ func hasImportPathPrefix(s, prefix string) bool {
 	return strings.HasPrefix(s, prefix+"/")
 }
 
+// itemizedFeedback itemizes resolved dependencies and logs the solution as
+// feedback.
+func itemizedFeedback(m *dep.Manifest, l *dep.Lock, pkgT pkgtree.PackageTree) {
+	// Algo:
+	// loop through lock projects
+	// 1. check in manifest for for same entry
+	// 2. if exists in manifest, it's direct dep, goes to manifest & lock
+	// 3. if not exists in manifest, check in pkgT
+	//	3.a. if exists in pkgT, it's direct dep, hint constraint, lock only
+	//	3.b. if not exists in pkgT, it's transitive dep, lock only
+	// 4. log feedback
+	for _, lock := range l.Projects() {
+		r, v, _ := gps.VersionComponentStrings(lock.Version())
+		projectPath := lock.Ident().ProjectRoot
+
+		cf := &internal.ConstraintFeedback{
+			Version:     v,
+			Revision:    r,
+			ProjectPath: projectPath,
+		}
+
+		if _, ok := m.Dependencies[projectPath]; ok {
+			// Direct dependency
+			cf.DependencyType = internal.DepTypeDirect
+			cf.ConstraintType = internal.ConsTypeConstraint
+		} else {
+			if _, ok := pkgT.Packages[string(projectPath)]; ok {
```

Comment:
The `Packages` map isn't quite what you're looking for - it's the list of packages within the current project, not the imports of those packages. What you want is a bit more elaborate. I've got code that does it over in #489 - https://github.com/sdboyer/dep/blob/1b8edb3436f39dbc29b4fae81dc1d51b2e77a303/cmd/dep/ensure.go#L390-L409 .